source/common/http/filter/buffer_filter.cc

```diff
@@ -38,10 +40,8 @@ FilterDataStatus BufferFilter::decodeData(Buffer::Instance&, bool end_stream) {
     return FilterDataStatus::Continue;
   } else if (callbacks_->decodingBuffer() &&
              callbacks_->decodingBuffer()->length() > config_->max_request_bytes_) {
-    // TODO(htuch): Switch this to Utility::sendLocalReply().
-    Http::HeaderMapPtr response_headers{new HeaderMapImpl{
-        {Headers::get().Status, std::to_string(enumToInt(Http::Code::PayloadTooLarge))}}};
-    callbacks_->encodeHeaders(std::move(response_headers), true);
+    Http::Utility::sendLocalReply(*callbacks_, stream_reset_, Http::Code::PayloadTooLarge,
```

Comment:
Vaguely wondering if we should have a `StreamDecoderFilterWithBoundedBuffer` class to inherit from that handles some of this automagically, but don't feel that strongly given how simple this watermark check is.